### PR TITLE
Implement guild-specific commands, manage permissions

### PR DESCRIPTION
**Pull request by davidvorona**
This is *really* rough, though seems to work. I'm torn on it to be honest, though the user experience is undoubtedly the best.

1. Adds a `privileged_roles` table that is modeled after the `tracking_channels` table
2. Adds methods for reading and writing privileged roles to/from the db
3. Adds a `privilegedRolesByGuild` map to state and the appropriate setters/getters
4. Initializes privileged roles data at the start of the app, much like `trackingChannels`
5. Adds a service `BearerTokenService` to get a bearer token from Discord unique to the application (uses the bot's client ID and secret) - see https://discord.com/developers/docs/topics/oauth2#client-credentials-grant
6. Adds a service `GuildCommandRolePermissionsManager` to overwrite a guild's command permissions using a bearer token
7. Adds a `guild` flag to `SlashCommand`
8. On startup, first the global (`command.guild !== true`) commands are registered w/o any changes
9. Afterwards, guild (`command.guild === true`) commands are registered for each guild
10. For each guild, if the guild has a saved privileged role, then a permission is added for each guild command that is also `privileged` using the new `GuildCommandRolePermissionsManager`

What still needs to be done:
1. `BearerTokenService` needs to be able to refresh the token if it expires and resend the request (currently, `/role` will begin to fail 1 day after the app starts unless the app is restarted) (COMPLETE)
2. Authorization/validation at the application-level needs to handle when a non-admin user with a privileged role uses a privileged guild command (currently it does not reject unauthorized usage); though this is unimportant, since authorization should be handled via native slash command behavior (INCOMPLETE, NOT NECESSARY)

Helpful info on permissions: https://discord.com/developers/docs/interactions/application-commands#permissions

There's a weird gotcha here that I think I understand. According to the link above, the authenticating user editing a command's permissions must have "permission to Manage Guild and Manage Roles in the guild where the command is being edited". This is why using the [Client Credential's grant](https://discord.com/developers/docs/interactions/application-commands#permissions) works, because the bot user is doing the authenticating and it can be given the `Manage Server` and `Manage Roles` permissions in its OAuth2 config. When it gets added to a server, it gets those permissions for that server. Like with guild commands, Discord does mention that you can use this grant type for development purposes, but I don't see why it also wouldn't work in production.  I still have not found anything suggesting we shouldn't.

Thoughts? Currently the code in this PR implements the best-possible UX:
1. Admins have access to `/track/clear/remove/channel`
2. If an admin uses `/role`, then the mentioned role immediately has access to `/track/clear/remove/channel`
3. All of this is natively supported slash command behavior, so no need to rely on application-level verification or messaging

### CODE REVIEW

File: src/commands.ts [hunk 1]
```diff
@@ -260,6 +267,33 @@ const slashCommands: SlashCommandsType = {
         },
         text: 'Show details of when each tracked player was last updated',
         privileged: true
+    },
+    role: {
+        options: [{
+            type: ApplicationCommandOptionType.Mentionable,
```
Comment (evanw555) [inline src/commands.ts:273]: This should be a role, since we're checking the role cache specifically

File: src/commands.ts [hunk 2]
```diff
@@ -260,6 +267,33 @@ const slashCommands: SlashCommandsType = {
...
+                await interaction.editReply(`${newPrivilegedRole} can now use **/track**, **/remove**, **/clear**, and **/channel**`);
+            } else {
+                await interaction.reply({
+                    content: 'Can only give a mentionable role (**@everyone**, **@Gamers**, etc.) privileges',
+                    ephemeral: true
+                });
+            }
+        },
+        text: 'Set a non-admin server role that can use /track, /remove, /clear, and /channel',
+        privileged: true
```
Comment (evanw555) [inline src/commands.ts:296]: How are we guaranteeing that only admins can use `/role`? If this is just `privileged`, then that means anyone with the privileged role can then themselves set the privileged role

File: src/pg-storage-client.ts [hunk 3]
```diff
@@ -185,6 +186,23 @@ export default class PGStorageClient {
     async writeBotCounter(userId: Snowflake, counter: number): Promise<void> {
         await this.client.query('INSERT INTO bot_counters VALUES ($1, $2) ON CONFLICT (user_id) DO UPDATE SET counter = EXCLUDED.counter;', [userId, counter]);
     }
+
+    async fetchAllPrivilegedRoles(): Promise<Record<Snowflake, Snowflake>> {
```
Comment (evanw555) [inline src/pg-storage-client.ts:190]: Add tests :)

File: src/pg-storage-client.ts [hunk 4]
```diff
@@ -185,6 +186,23 @@ export default class PGStorageClient {
...
+        const result: Record<Snowflake, Snowflake> = {};
+        const queryResult = await this.client.query<{guild_id: Snowflake, role_id: Snowflake}>('SELECT * FROM privileged_roles;');
+        for (const row of queryResult.rows) {
+            result[row.guild_id] = row.role_id;
+        }
+        return result;
+    }
+    
+    async deletePrivilegedRole(guildId: Snowflake, roleId: Snowflake): Promise<void> {
```
Comment (evanw555) [inline src/pg-storage-client.ts:199]: Looks like this isn't used now. Are we going to use this when being removed from guilds? Perhaps later we can implement some `deleteAllGuildData` method

File: src/util.ts [hunk 5]
```diff
@@ -495,3 +498,110 @@ export function getNextFridayEvening(): Date {
     }
     return nextFriday;
 }
+
+class BearerTokenService {
```
Comment (evanw555) [inline src/util.ts:502]: Put this in its own file

File: src/util.ts [hunk 6]
```diff
@@ -495,3 +498,110 @@ export function getNextFridayEvening(): Date {
...
+            }
+            logger.log(`There was an error refreshing the bearerToken: ${err}`, MultiLoggerLevel.Error);
+            throw err;
+        }
+    }
+}
+
+// TODO: This service needs to call to refresh the bearerToken when the API returns
+// a token expired response.
+export class GuildCommandRolePermissionsManager {
```
Comment (evanw555) [inline src/util.ts:547]: Put this in its own file

File: src/util.ts [hunk 7]
```diff
@@ -495,3 +498,110 @@ export function getNextFridayEvening(): Date {
...
+                }
+                // Set the new permissions for the command
+                return guild.commands.permissions.add(commandPermissions);
+            }));
+            logger.log(`Updated guild '${guild.id}' command permissions: \`${JSON.stringify(results)}\``, MultiLoggerLevel.Debug);
+        }
+    }
+}
+
+export const guildCommandRolePermissionsManager = new GuildCommandRolePermissionsManager();
```
Comment (evanw555) [inline src/util.ts:607]: All singletons should be exported from their own files in `src/instances/*.ts`

File: src/command-handler.ts [hunk 8]
```diff
@@ -128,18 +174,20 @@ class CommandHandler {
...
         const debugString = `\`${interaction.user.tag}\` executed command \`${interaction.toString()}\` in ${interaction.channel}`;
         try {
             if (command.failIfDisabled) {
                 CommandHandler.failIfDisabled();
             }
             if (command.privileged) {
-                CommandHandler.assertIsAdmin(interaction);
+                if (command.guild) {
+                    // TODO: This code needs to check that the interacting user has the role that is granted privilege.
+                } else CommandHandler.assertIsAdmin(interaction);
```
Comment (evanw555) [inline src/command-handler.ts:190]: nit: use braces

File: src/commands.ts (hunk 2 above)
Comment (evanw555) [reply src/commands.ts:296]: Is that what the logic with the TODO does?

**Review by evanw555 — COMMENTED** [review]
Overall, this code needs a lot more comments, it's hard to follow a lot of the command logic

File: src/commands.ts (hunk 2 above)
Comment (davidvorona) [reply src/commands.ts:296]: No, because /role is not a `guild` command.
